### ISSUE TITLE
resin local push should use the `--cache-from` for builds if available

### ISSUE THREAD

**Issue by shaunmulligan** [resin-io/resin-cli#605]
In newer versions of docker, the it is possible to specify a images on the target device to consider for cache. In a local build. The build step should consider any of the images that already exist on the device to be sources for build cache. Here is the reference for docker build: https://docs.docker.com/edge/engine/reference/commandline/build/#description

This will drastically speed up the first local push if someone has already do a git push to the device.

**Comment by shaunmulligan** [comment]
@pimterry do you have a sense of how difficult this is to do? it should make the first local build much faster if it can use the already pushed image on the device.

**Comment by pimterry** [comment]
@shaunmulligan not off the top of my head, no. I wouldn't have thought it'll be too tough, but I haven't really delved into the docker interactions within local push much yet. I'll have a look later today to investigate.

**Comment by shaunmulligan** [comment]
cool, that would be great :)

**Comment by pimterry** [comment]
@shaunmulligan I've had a go at doing this manually (pointing my local docker client at a live device) and I can't get it to work. Steps:

* I git push a repo to the application
* That builds and starts successfully on the device
* I set my `$DOCKER_HOST` to point to the device locally
* If I run `docker ps` I can see the deployed running container
* If I run `docker build . --cache-from <app image id>` from the same source directory, it pulls the base image and runs a full normal build, without any apparent caching.

Device is a pi3 running 2.0.6+rev3.dev.

I'm not exactly sure how the caching mechanisms work in here, but one suspicious note I've found looking into this is that while `docker history <supervisor image id>` comes back with all the layers for the supervisor, `docker history <app image id>` comes back with nothing. The image size in `docker images` is shown as only 17B, no idea what's going on there. Are we dropping all the image contents entirely somewhere?

As a test, I tried manually pulling the full app image onto the device (`docker pull registry2.resin.io/...`), and `docker images` and `docker history` now show broadly what I expect (though all but the top layer of history are `<missing>`), but `--cache-from` still doesn't seem to do anything (both normal and `cache-from` builds now skip downloading the base image, before running the rest of the build uncached).

I think the issue is that we don't have any of the intermediate layers on the device - it's not just that we need the resulting previous image to exist, the cached layers we really need are the intermediate layers from the previous build, and since there've been no builds on this device, and we don't distribute those intermediate layers, that's not going to work.

Does that sound plausible? I am missing something that should make this work? I think we could potentially save intermediate layers in the builder, and distribute them on demand to allow for this, but that's a larger problem and afaict we don't currently do any of that.

**Comment by shaunmulligan** [comment]
Hey @pimterry , off the top of my head I can think of two things that might have tripped you up. The first is that resinOS 2.0.6 has docker 1.10.3 and I think the cache-from flag was only introduced later.

The second thing is that if your docker image is showing as `17B` it's probably because you have deltas enabled and deltas in their current form do not preserve the layer history (the new Pepe implementation for hostapps does)

**Comment by pimterry** [comment]
@shaunmulligan good catch, thanks! Both of things were true. I've now moved to a device with 2.3.0+rev1.dev, which is Docker 17.03.1-ce, which definitely includes `cache-from`, and I've disabled deltas.

It now seems to use the cache for docker steps correctly, but does still do the initial base image download from scratch first, I think due to https://github.com/moby/moby/issues/31613.

That's fine for now though, as it still saves quite a bit of time. Should be pretty easy to integrate from here I think, just need to get all the image ids already on the device, and add a cache-from during build for each of them. Barring any problems with Dockerode support for this, it should be quick - I'll try and have a first pass at it tomorrow.
